pkg/cloudprovider/openstack/openstack.go

```diff
@@ -625,28 +626,41 @@ func (lb *LoadBalancer) UpdateTCPLoadBalancer(name, region string, hosts []strin
 func (lb *LoadBalancer) DeleteTCPLoadBalancer(name, region string) error {
 	glog.V(4).Infof("DeleteTCPLoadBalancer(%v, %v)", name, region)
 
-	vip, err := getVipByName(lb.network, name)
-	if err != nil {
-		return err
+	vip, vipErr := getVipByName(lb.network, name)
+	vipExists := true
+	if vipErr != nil {
+		if vipErr != ErrNotFound {
+			return vipErr
+		}
+		vipExists = false
 	}
 
-	pool, err := pools.Get(lb.network, vip.PoolID).Extract()
-	if err != nil {
-		return err
+	pool, poolErr := pools.Get(lb.network, vip.PoolID).Extract()
+	poolExists := true
+	if poolErr != nil {
+		detailedErr, ok := poolErr.(*gophercloud.UnexpectedResponseCodeError)
+		if !ok || detailedErr.Actual != http.StatusNotFound {
+			return poolErr
+		}
+		poolExists = false
 	}
 
 	// Have to delete VIP before pool can be deleted
-	err = vips.Delete(lb.network, vip.ID).ExtractErr()
-	if err != nil {
-		return err
+	if vipExists {
+		err := vips.Delete(lb.network, vip.ID).ExtractErr()
+		if err != nil {
+			return err
+		}
 	}
 
 	// Ignore errors for everything following here
 
-	for _, monId := range pool.MonitorIDs {
-		pools.DisassociateMonitor(lb.network, pool.ID, monId)
+	if poolExists {
+		for _, monId := range pool.MonitorIDs {
+			pools.DisassociateMonitor(lb.network, pool.ID, monId)
+		}
+		pools.Delete(lb.network, pool.ID)
```

Comment:
Since a couple weeks ago, load balancers have been named using their service's UID, so there will never be name conflicts that enable reusing old pools. Given that, would you be ok with switching over to getting the pool by name? This would make the OpenStack behavior almost identical to the GCE behavior, meaning it'll be less likely to break in weird ways in the future.

In the meantime, I can have this function gracefully return nil if the vip doesn't exist. Once you've switched the pool lookups to be by name, we can go back to trying to delete the pool even if the vip has already been deleted.

How does that sound? I'll open up an issue for doing this work and for cleaning up the monitors that aren't currently being deleted.